References
----------

.. [Bergen20] Bergen, V. *et al.* (2020),
   *Generalizing RNA velocity to transient cell states through dynamical modeling*,
   `Nature Biotechnology <https://doi.org/10.1038/s41587-020-0591-3>`__.

.. [Coifman05] Coifman *et al.* (2005),
    *Geometric diffusions as a tool for harmonic analysis and structure definition of data: Diffusion maps*,
    `PNAS <https://doi.org/10.1073/pnas.0500334102>`__.

.. [GPCCA18] Reuter, B. *et al.* (2018),
   *Generalized Markov State Modeling Method for Nonequilibrium Biomolecular Dynamics: Exemplified on Amyloid β
   Conformational Dynamics Driven by an Oscillating Electric Field.*,
   `Journal of Chemical Theory and Computation <https://doi.org/10.1021/acs.jctc.8b00079>`__.

.. [Haghverdi16] Haghverdi *et al.* (2016),
   *Diffusion pseudotime robustly reconstructs branching cellular lineages*,
   `Nature Methods <https://doi.org/10.1038/nmeth.3971>`__.

.. [Li2020] Li, T. *et al.* (2020),
    *On the Mathematics of RNA Velocity I: Theoretical Analysis*,
    `bioRxiv <https://doi.org/10.1101/2020.09.19.304584>`__.

.. [Lung20] Strunz, M. *et al.* (2020),
    *Alveolar regeneration through a Krt8+ transitional stem cell state that persists in human lung fibrosis*,
    `Nature Communications <https://doi.org/10.1038/s41467-020-17358-3>`__.

.. [Macosko15] Macosko, E. Z. *et al.*, (2015),
    *Highly Parallel Genome-wide Expression Profiling of Individual Cells Using Nanoliter Droplets*,
    `Cell <https://doi.org/10.1016/j.cell.2015.05.002>`__.

.. [MAGIC18] Van Dijk, D. *et al.* (2018),
    *Recovering Gene Interactions from Single-Cell Data Using Data Diffusion*,
    `Cell <https://doi.org/10.1016/j.cell.2018.05.061>`__.

.. [Manno18] La Manno *et al.* (2018),
    *RNA velocity of single cells*,
    `Nature <https://doi.org/10.1038/s41586-018-0414-6>`__.

.. [Morris18] Biddy, B.A., Kong, W., Kamimoto, K. *et al.* (2018),
    *Single-cell mapping of lineage and identity in direct reprogramming.*,
    `Nature 564, 219–224 <https://doi.org/10.1038/s41586-018-0744-4>`__.

.. [Panc19] Bastidas-Ponce *et al.* (2019),
    *Comprehensive single cell mRNA profiling reveals a detailed roadmap for pancreatic endocrinogenesis*,
    `Development <https://doi.org/10.1242/dev.173849>`__.

.. [Reuter19] Reuter, B. *et al.* (2019),
    *Generalized Markov modeling of nonreversible molecular kinetics*,
    `The Journal of Chemical Physics <https://doi.org/10.1063/1.5064530>`__.

.. [Robinson10] Robinson, M. D. *et al.* (2010),
    *A scaling normalization method for differential expression analysis of RNA-seq data*,
    `Genome Biology <https://doi.org/10.1186/gb-2010-11-3-r25>`__.

.. [Setty19] Setty *et al.* (2019),
    *Characterization of cell fate probabilities in single-cell data with Palantir*,
    `Nature Biotechnology <https://doi.org/10.1038/s41587-019-0068-4>`__.

.. [Velten17] Velten, L. *et al.* (2017),
    *Human haematopoietic stem cell lineage commitment is a continuous process.*,
    `Nat Cell Biol 19, 271–281 <https://doi.org/10.1038/ncb3493>`__.

.. [Tolver16] Tolver (2016),
    *An introduction to Markov chains*,
    `Recuperado el 15 <http://old.math.ku.dk/noter/filer/stoknoter.pdf>`__.

.. [Weinreb18] Weinreb *et al.* (2018),
    *Fundamental limits on dynamic inference from single-cell snapshots*,
    `PNAS <https://doi.org/10.1073/pnas.1714723115>`__.

.. [Wolf18] Wolf *et al.* (2018),
    *Scanpy: large-scale single-cell gene expression data analysis*,
    `Genome Biology <https://doi.org/10.1186/s13059-017-1382-0>`__.

.. [Wolf19] Wolf *et al.* (2019),
    *PAGA: Graph abstraction reconciles clustering with trajectory inference through a topology preserving map of
    single cells*,
    `Genome Biology <https://doi.org/10.1186/s13059-019-1663-x>`__.
